packages/flutter/lib/src/material/navigation_bar.dart

```diff
@@ -0,0 +1,1204 @@
+    final ColorScheme colorScheme = theme.colorScheme;
+    final NavigationBarThemeData navigationBarTheme = NavigationBarTheme.of(context);
+    final Animation<double> animation = NavigationDestinationInfo.of(context).selectedAnimation;
+
+    return NavigationDestinationBuilder(
+      label: label,
+      tooltip: tooltip,
+      buildIcon: (BuildContext context) {
+        final Widget selectedIconWidget = IconTheme.merge(
+          data: navigationBarTheme.selectedIconTheme ?? IconThemeData(
+            size: 24,
+            color: colorScheme.onSurface,
+          ),
+          child: selectedIcon ?? icon,
+        );
+        final Widget unselectedIconWidget = IconTheme.merge(
+          data: navigationBarTheme.unselectedIconTheme ?? IconThemeData(
+            size: 24,
+            color: colorScheme.onSurface,
+          ),
+          child: icon,
+        );
+
+        return Stack(
+          alignment: Alignment.center,
+          children: <Widget>[
+            NavigationIndicator(animation: animation),
+            _StatusTransitionWidgetBuilder(
+              animation: animation,
+              builder: (BuildContext context, Widget? child) {
+                return animation.isForwardOrCompleted
+                    ? selectedIconWidget
+                    : unselectedIconWidget;
+              },
+            ),
+          ],
+        );
+      },
+      buildLabel: (BuildContext context) {
+        final TextStyle? defaultTextStyle = theme.textTheme.overline?.copyWith(
+          color: colorScheme.onSurface,
+        );
+        final TextStyle? effectiveSelectedLabelTextStyle = navigationBarTheme.selectedLabelTextStyle ?? defaultTextStyle;
+        final TextStyle? effectiveUnselectedLabelTextStyle = navigationBarTheme.unselectedLabelTextStyle ?? defaultTextStyle;
+        return Padding(
+          padding: const EdgeInsets.only(top: 4),
+          child: _ClampTextScaleFactor(
+            // Don't scale labels of destinations, instead, tooltip text will
+            // upscale.
+            upperLimit: 1,
+            child: Text(
+              label,
+              style: animation.isForwardOrCompleted
+                  ? effectiveSelectedLabelTextStyle
+                  : effectiveUnselectedLabelTextStyle,
+            ),
+          ),
+        );
+      },
+    );
+  }
+}
+
+/// Widget that handles the semantics and layout of a navigation bar
+/// destination.
+///
+/// Prefer [NavigationDestination] over this widget, as it is a simpler
+/// (although less customizable) way to get navigation bar destinations.
+///
+/// The icon and label of this destination are built with [buildIcon] and
+/// [buildLabel]. They should build the unselected and selected icon and label
+/// according to [NavigationDestinationInfo.selectedAnimation], where 0 is
+/// unselected and 1 is selected.
+///
+/// See [NavigationDestination] for a example.
+class NavigationDestinationBuilder extends StatelessWidget {
```

Comment:
This building block would allow you to provide more complex destinations (for example an icon that animates when it is tapped) while still keeping the semantics and tooltip behavior you expect